src/main/java/com/redhat/rhjmc/containerjfr/net/web/WebServer.java

```diff
@@ -0,0 +1,327 @@
+                    Duration.between(start, Instant.now()).toMillis()
+            )));
+            router.handle(req);
+        });
+    }
+
+    public void stop() {
+        this.server.requestHandler(null);
+
+        recordings.clear();
+        downloadCounts.clear();
+    }
+
+    public void addRecording(IRecordingDescriptor descriptor) {
+        recordings.put(descriptor.getName(), descriptor);
+        downloadCounts.put(descriptor.getName(), 0);
+    }
+
+    public void removeRecording(IRecordingDescriptor descriptor) {
+        recordings.remove(descriptor.getName());
+        downloadCounts.remove(descriptor.getName());
+    }
+
+    public int getDownloadCount(String recordingName) {
+        return this.downloadCounts.getOrDefault(recordingName, -1);
+    }
+
+    public URL getHostUrl() throws UnknownHostException, MalformedURLException, SocketException {
+        return new URL("http", netConf.getWebServerHost(), netConf.getExternalWebServerPort(), "");
+    }
+
+    public String getDownloadURL(String recordingName)
+            throws UnknownHostException, MalformedURLException, SocketException {
+        return String.format("%s/recordings/%s", this.getHostUrl(), recordingName);
+    }
+
+    public String getReportURL(String recordingName)
+            throws UnknownHostException, MalformedURLException, SocketException {
+        return String.format("%s/reports/%s", this.getHostUrl(), recordingName);
+    }
+
+    private Optional<InputStream> getRecordingInputStream(String recordingName) throws FlightRecorderException {
+        if (recordings.containsKey(recordingName)) {
+            return Optional.of(service.openStream(recordings.get(recordingName), false));
+        }
+        try {
+            Optional<Path> savedRecording = Files.list(savedRecordingsPath)
+                    .filter(saved -> saved.getFileName().toFile().getName().equals(recordingName) || saved.getFileName().toFile().getName().equals(recordingName + ".jfr"))
+                    .findFirst();
+            if (savedRecording.isPresent()) {
+                return Optional.of(Files.newInputStream(savedRecording.get(), StandardOpenOption.READ));
+            }
+        } catch (Exception e) {
+            logger.error(e);
+        }
+        return Optional.empty();
+    }
+
+    private void endWithJsonKeyValue(String key, String value, HttpServerResponse response) {
+        response.end(String.format("{\"%s\":\"%s\"}", key, value));
+    }
+
+    private HttpServerResponse writeInputStream(InputStream inputStream, HttpServerResponse response) throws IOException {
+        // blocking function, must be called from a blocking handler
+        byte[] buff = new byte[WRITE_BUFFER_SIZE]; // 64 KB
+        int n;
+        while (true) {
+            n = inputStream.read(buff);
+            if (n == -1) {
+                break;
+            }
+
+            response.write(Buffer.buffer().appendBytes(buff, 0, n));
```

Comment:
I'm not familiar with vertx's stream writing pieces, but this looks suspicious. Why is a new buffer (`Buffer.buffer()`) being allocated on each iteration of the loop?

And, is there no way to make this non-blocking since the source is a stream?

Reply:
Buffer allocation will be moved outside the loop in the next commit.

The source is a stream, but a `java.io` blocking stream. `read()` calls are guaranteed to return immediately. Although there is a way to convert a blocking `InputStream` into a `java.nio` async `ReadableByteChannel`, but such a method conversion cannot utilize epoll/select/kqueue and signals and achieve true asynchronous, but instead creating a worker thread that still calls blocking functions. (I'm not 100% sure about this.) Additionally, since the report generator is a synchronous API anyway, I chose the straight forward way of using blocking handlers.

Reply:
This is used when the client attempts to download a recording file though, isn't it? If it's made asynchronous, even if backed by a blocking stream, it will allow for data to flow out over the network and be in-flight as the rest of the recording is loaded. It would also reduce the memory footprint of sending out the recording, since it wouldn't need to be copied entirely into a buffer before being handed off to the network stack.

Reply:
If that's the concern, the data is already being streamed to clients with HTTP chunked transfer encoding. Every call to `.write` will queue a 64 KB chunk (in FIFO order of course) to send on the next appropriate event loop tick.

And this reminds me why I was keeping allocating `Buffer.buffer()` in the loop: if the content of a buffer is rewritten with the next chunk before it is sent out, the file client receives will be corrupted. So there must be a new buffer allocated **inside the loop** for every new chunk.

Reply:
Or I can code it so it only reads a new chunk after the last chunk is sent out. In this case only one 64 KB buffer is needed.

Reply:
I see - it seems odd that the API uses the `.write` calls to also delineate the chunks within the write queue, but that is what the docs say so I suppose it's how it's done.

Using a single buffer and only refilling it after the previous contents have been flushed will give us a constant memory size footprint, which is very nice, but also has the potential to reduce the network throughput (although also lowering initial response latency). Maybe this is worth trying out. Or, could you implement both and make the selection configurable by a property or env var? I think we'll want to test both solutions in the real-world for a while and see what works best. It may even be something we leave for users to decide.

Reply:
Also, while looking through the vertx API, I noticed the `.sendFile` method. Since the requested recording can be either in-memory or on-disk, this is something that we can/should take advantage of. That information about whether the recording is in-memory or on-disk is currently lost by the `getRecordingInputStream` implementation, but I think it's worth refactoring things so that this distinction is preserved and the appropriate API used. This would make the download of on-disk recordings *way* more efficient, so the cost vs. benefit analysis of how we treat the buffers in the discussion above would only apply to in-memory recordings.

In the future it might even make sense to drop the ability to directly download files out of memory without first flushing the recording to disk storage.

Reply:
No problem. This will be implemented in the next commit.

Reply:
I have implemented a synchronous method writing an `InputStream`. It can be enabled by defining `USE_SYNC_WRITE_INPUT_STREAM` environment variable, regardless of the value.

`.sendFile()` is very efficient because the copying of the content between two fds is directly handled by the kernel. The actual content doesn't route around to userspace. However, it's worth noting this advantage doesn't apply if we enable end-to-end SSL encryption in the future, as the encryption process still needs to be done in userspace.

Reply:
I have implemented a synchronous method writing an InputStream. It can be enabled by defining `USE_SYNC_WRITE_INPUT_STREAM` environment variable, regardless of the value.

`.sendFile()` is very efficient because the copying of the content between two fds is directly handled by the kernel. The actual content doesn't route around to userspace. However, it's worth noting this advantage doesn't apply if we enable end-to-end SSL encryption in the future, as the encryption process still needs to be done in userspace.

Reply:
That's a good point about SSL. It will be something to take into consideration since flight recordings can easily contain sensitive information.

Reply:
I think there's still some optimization to be done to make the response stream the data better using a vertx "Pump", but this is an optimization we can look at later. I'll mark this as resolved for now and we can file an issue for exploring performance optimization.